pkg/cloudprovider/providers/azure/azure_blobDiskController.go

```diff
@@ -0,0 +1,1061 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/golang/glog"
+	"github.com/rubiojr/go-vhd/vhd"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+type storageAccountState struct {
+	name                    string
+	saType                  string
+	key                     string
+	diskCount               int32
+	isValidating            int32
+	defaultContainerCreated bool
+}
+
+//BlobDiskController : blob disk controller struct
+type BlobDiskController struct {
+	common   *controllerCommon
+	accounts map[string]*storageAccountState
+}
+
+var defaultContainerName = ""
+var storageAccountNamePrefix = ""
+var storageAccountNameMatch = ""
+var initFlag int64
+
+var accountsLock = &sync.Mutex{}
+
+func newBlobDiskController(common *controllerCommon) (*BlobDiskController, error) {
+	c := BlobDiskController{common: common}
+	err := c.init()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &c, nil
+}
+
+//AttachBlobDisk : attaches a disk to node and return lun # as string
+func (c *BlobDiskController) AttachBlobDisk(nodeName string, diskURI string, cacheMode string) (int, error) {
+	// K8s in case of existing pods evication, will automatically attepmt to attach volumes
+	// to a different node. Though it *knows* which disk attached to which node.
+	// the following guards against this behaviour
+
+	// avoid:
+	// Azure in case of blob disks, does not maintain a list of vhd:attached-to:node
+	// The call  attach-to will fail after it was OK on the ARM VM endpoint
+	// possibly putting the entire VM in *failed* state
+	noLease, e := c.diskHasNoLease(diskURI)
```

Comment:
This is yet another race condition: just because there is no lease on the disk doesn't mean somebody else wants to attach the disk before your attach request.